confluent_kafka/producer.py

```diff
@@ -0,0 +1,139 @@
+                 "all keyword arguments must match the constructor signature explicitly.",
+                 category=DeprecationWarning, stacklevel=2)
+
+        # Ensure the default serializer cannot be overwritten with None on instantiation
+        if key_serializer is None:
+            key_serializer = SerializingProducer.byteSerializer
+
+        if value_serializer is None:
+            value_serializer = SerializingProducer.byteSerializer
+
+        self._key_serializer = key_serializer
+        self._value_serializer = value_serializer
+
+        # Callbacks can be set in the conf dict or *ideally* as parameters.
+        # Raise a SyntaxError if a callback is set in both places.
+        for var, name in [(logger, 'logger'), (error_cb, 'error_cb'),
+                          (stats_cb, 'stats_cb'), (throttle_cb, 'throttle_cb')]:
+            if all([var, conf.get(name, None)]):
+                raise SyntaxError("{} parameter repeated".format(name))
+            if var is None:
+                var = conf.get(name, None)
+
+        super(SerializingProducer, self).__init__(conf, error_cb=error_cb, stats_cb=stats_cb,
+                                                  throttle_cb=throttle_cb, logger=logger)
+
+    def produce(self, topic, value=None, key=None, partition=PARTITION_UA,
+                on_delivery=None, callback=None, timestamp=0, headers=None):
+        """
+            Produces message to Kafka
+
+            :param str topic: Topic to produce message to.
+            :param str|bytes|obj value: Message payload; value_serializer required for non-character values.
+            :param str|bytes|obj key: Message key payload; key_serializer required for non-character keys.
+            :param int partition: Partition to produce to, else uses the configured built-in partitioner.
+                Default value is PARTITION_UA(round-robin over all partitions).
+            :param func on_delivery(confluent_kafka.KafkaError, confluent_kafka.Message):
+                Delivery report callback to call on successful or failed delivery.
+            :param func callback(confluent_kafka.KafkaError, confluent_kafka.Message):
+                See on_delivery.
+            :param int timestamp: Message timestamp (CreateTime) in milliseconds since epoch UTC
+                (requires librdkafka >= v0.9.4, api.version.request=true, and broker >= 0.10.0.0).
+                Default value is current time.
+            :param dict|list headers: Message headers to set on the message. The header key must be a string while
+                the value must be binary, unicode or None. Accepts a list of (key,value) or a dict.
+                (Requires librdkafka >= v0.11.4 and broker version >= 0.11.0.0)
+        """
+
+        super(SerializingProducer, self).produce(topic, self._value_serializer(topic, value),
+                                                 self._key_serializer(topic, key), partition, on_delivery=on_delivery,
```

Comment:
I think it is a bad idea to hold on to the original key/value object (pre-serialization) since it will keep them from being garbage collected.
A user can bind the on_delivery callback to their original object to have it available in the delivery report callback.

Reply:
GC aside I personally feel that violates the implicit contract that comes with the delivery report. The way I interpret it the DR is meant to tell me what was produced not what I meant to produce. So the counter argument is that this gives me quick validation that my payload was serialized as I had intended. If I need verification it was serialized correctly I can use a lambda to carry state with it. Otherwise I have no way of telling if it was actually serialized unless I consume the data back down.

Reply:
'quick validation that my payload was serialized as I had intended' - it seems like you don't trust python or your serialization format of choice? Can I suggest you get a new language or serialization format of choice? No one should want/need to do this. Also, if they did for some reason, it is possible for them to drive the API such that they can (but that should not be the easy path).

Re: additional memory use: there should be a config switch to allow this to be turned off for people who don't need it / don't want to pay for the memory overhead. In C# I have it on by default so people don't get confused with null Key / Value fields (i.e. cater for beginner by default), but in python you could have it off by default since you would just not provide the fields.

Re: binding this to the on_delivery callback: you're right, but users couldn't use a static callback method. If the user does want state maintained, it will very often be unserialized message data, so it's is a nice feature to provide this out of the box, allowing for simpler usage patterns in the common case.

If you do want to expose the serialized bytes and not the original data (i think this is not a good idea), how about using a different name, to indicate this is the serialized data.

Reply:
The problem is that it will hold on to any resources referenced by the object, say a database response, a file, or something else that is costly.
If you produce 1 gig of data with this approach, your application will, worst case, peak out at 2 gigs of memory usage just for the original objects and the copied message payloads.


As for validating the serialized data, I agree with Matt that's not valuable.

Reply:
holding onto additional resources would certainly be an important (dominating) consideration if that were likely, but I don't think it is. I'd be convinced if i saw a concrete example of where you think someone might accidentally do this.

re memory usage, yes, which is why you need a flag to turn it off (or a flag to turn it on for those who want the feature - simply not providing the property in python is an option - in statically typed languages it's not, hence my preference for it to be on by default there to avoid confusion). that said, i'd also be inclined to err on the side of non-power-user in python as well.

Reply:
Right I'm not saying serialization verification is the primary motivator here just that you can still do that even without returning the message in its original form. Instead I was referring to the fact that if needed you can absolutely see both forms of the message with minimal effort. A `delivery report` in my eyes should be reporting on the actual message produced. Which in this case is the post-serialized form. Either way providing a lambda as a callback requires a marginal amount of additional effort over providing a configuration flag.

Reply:
I would argue that with a serializer the key and value of a message isn't of much interest in the delivery report, the original object is.

Agree that a lambda or object-bound method is sufficient to pass state between produce and delivery report.

Reply:
When your writing code in a scripting fashion, as you are in python, quick indicators of success are handy. If I send `hello world` as  as a string that I wanted to be base64 encoded it's nice to have some sort of visual indication that it worked. I'm not saying this is the right way to approach things but it's how a lot of people make incremental progress with more complex bash/python scripts. You write some code, you run it, you sanity check the output, you write some more code.

Reply:
That doesn't belong here, you do that verification on the serializer.

Reply:
That' fine, as I said it may not be right but it happens. I also agree with the GC implications either way so I think for the time being this particular issue is resolved.

Reply:
to be concrete, to get the original k/v in the delivery report, I think the simplest you could do is:

```
if __name__ == '__main__':
    ....
    for line in sys.stdin:
        def line_cb_creator(l):
            def line_cb(err, msg):
                sys.stderr.write('delivered: %s' % l)
            return line_cb
        p.produce(topic, line.rstrip(), callback=line_cb_creator(line))
        p.poll(0)
```

the lexical scope of the loop parameter in python here is the main function, which is why the `line_cb_creator` function is necessary. I think that's too hard for an 'easy to use' python api, and think that at least ultimately we would want to plan to be adding the original values to the delivery report, and so need to make sure the behavior introduced now doesn't conflict with the foreseen API. we include serialized data in `key()` and `value()` in the existing api. I think that means that to avoid making breaking changes, we need to overload those methods now to return the unserialized data.

Reply:
For now we are going to move forward with leaving the delivery report key/value in their serialized form.